src/webpub_manifest_parser/core/ast.py

```diff
@@ -0,0 +1,831 @@
+
+
+@six.add_metaclass(ABCMeta)
+class Visitable(object):
+    """Interface for objects walkable by AST visitors."""
+
+    @abstractmethod
+    def accept(self, visitor):
+        """Accept  the specified visitor.
+
+        :param visitor: Visitor object
+        :type visitor: Visitor
+        """
+        raise NotImplementedError()
+
+
+@six.add_metaclass(ABCMeta)
+class Extendable(object):
+    """Abstract class adding ability to extend classes.
+
+    For example, RWPM link properties can be extended by EPUB link properties and OPDS 2.0 link properties.
+    """
+
+    extensions = None
+
+    @classmethod
+    def get_extension(cls):
+        """Return a new class having all extensions as mixins.
+
+        :return: New class containing extensions as mixins.
+        :rtype: Type
+        """
+        if not cls.extensions:
+            return cls
+
+        if len(cls.extensions) == 1:
+            return cls.extensions[0]
+
+        class_names = [cls.__name__] + [
+            extension.__name__ for extension in cls.extensions
+        ]
+        extended_class_name = "_".join(class_names)
+        extended_class = type(extended_class_name, tuple(cls.extensions), {})
+
+        return extended_class
+
+
+class Node(PropertiesGrouping, Visitable, Extendable):
+    """Base class for all AST nodes."""
+
+    def accept(self, visitor):
+        """Accept the specified visitor.
+
+        :param visitor: Visitor object
+        :type visitor: Visitor
+        """
+        visitor.visit(self)
+
+
+class LinkProperties(Node):
+    """Link properties."""
+
+    clipped = BooleanProperty("clipped", required=False)
+    fit = EnumProperty(
+        "fit", required=False, items=["contain", "cover", "width", "height"]
+    )
+    orientation = EnumProperty(
+        "orientation", required=False, items=["auto", "landscape", "portrait"]
+    )
+    page = EnumProperty("page", required=False, items=["left", "right", "center"])
+    spread = EnumProperty(
+        "spread", required=False, items=["auto", "both", "none", "landscape"]
+    )
+
+    def __hash__(self):
+        """Calculate the hash.
+
+        :return: Hash
+        :rtype: int
+        """
+        return hash((self.clipped, self.orientation, self.page, self.spread))
+
+    def __repr__(self):
+        """Return a string representation of the object.
+
+        :return: String representation
+        :rtype: str
+        """
+        return u"<LinkProperties(clipped={0}, fit={1}, orientation={2}, page={3}, spread={4})>".format(
+            self.clipped, self.fit, self.orientation, self.page, self.spread
+        )
+
+
+class Link(Node):
+    """Link to another resource."""
+
+    href = URITemplateProperty("href", required=True)
+    templated = BooleanProperty("templated", required=False)
+    type = StringProperty("type", required=False)
+    title = StringProperty("title", required=False)
+    rel = ArrayOfStringsProperty("rel", required=False)
```

Comment:
For `rel`, using a set would be more suitable instead of an array.

Also we tend to pluralize these array properties compared to the raw RWPM, so `rels`, `alternates` and `languages`.

Reply:
Good catch, thank you! `ArrayOfStringsProperty` has optional parameter `unique_items` which can be used to enforce uniqueness of items but I think I'll better create a new property `SetOfStringsProperty`.